scenario/mocking.py

```diff
@@ -69,26 +70,51 @@ class ActionMissingFromContextError(Exception):
 
 
 class _MockExecProcess:
-    def __init__(self, command: Tuple[str, ...], change_id: int, out: "ExecOutput"):
+    def __init__(
+        self,
+        command: Tuple[str, ...],
+        change_id: int,
+        exec: "Exec",
+        container: "ContainerSpec",
+    ):
         self._command = command
         self._change_id = change_id
-        self._out = out
+        self._exec = exec
         self._waited = False
-        self.stdout = StringIO(self._out.stdout)
-        self.stderr = StringIO(self._out.stderr)
+        self.stdout = StringIO(self._exec.stdout)
+        self.stderr = StringIO(self._exec.stderr)
+        self.stdin = StringIO(self._exec.stdin)
+        self._container = container
+
+    def _store_stdin(self):
+        execs = {
+            e
+            for e in self._container.execs
+            if e.command_prefix != self._exec.command_prefix
+        }
+        exec = dataclasses.replace(self._exec, stdin=self.stdin.read())
+        execs.add(exec)
+        self._container._update_execs(execs)
 
     def wait(self):
         self._waited = True
-        exit_code = self._out.return_code
+        self._store_stdin()
+        exit_code = self._exec.return_code
```

Comment:
Is it so that the API doesn't allow testing charms that issue failing commands on purpose, like `stat /no-such-file` ?

Reply:
Testing commands that fail should work (and did before, so if they have stopped working then I broke something and need to fix it). You just provide a non-zero exit code.